### ISSUE TITLE
New EXX implementation not working with gcc+openmpi+openblas

### ISSUE THREAD

**Issue by xinyangd** [deepmodeling/abacus-develop#1643]
### Describe the bug

1. When running with more than one mpi process, the program would get stuck after initialization.
2. When running with single mpi process, the program sometimes produces wrong results. i.e. with the hse example in the example folder, when changing the number of k points in KPT file to 111, Gamma_only=0 and Gamma_only=1 would give different results.

### Environment

gcc-11, openmpi, openblas

**Comment by xinyangd** [comment]
Here are the three output files:
[outfile.zip](https://github.com/deepmodeling/abacus-develop/files/10166612/outfile.zip)

**Comment by xinyangd** [comment]
Peize Lin will work on this.

**Comment by caic99** [comment]
> 1. When running with more than one mpi process, the program would get stuck after initialization.

I'm not able to reproduce this error based on MPICH. This problem might be related to OpenMPI.

**Comment by caic99** [comment]
I'm trying to run this hse example with sanitizer, but failed. Maybe we should fix this error first.

```
Initial plane wave basis and FFT box
 ---------------------------------------------------------
=================================================================
==56190==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x612000041ba8 at pc 0x55e0add9464d bp 0x7ffc5de0e610 sp 0x7ffc5de0e600
READ of size 8 at 0x612000041ba8 thread T0
    #0 0x55e0add9464c in std::vector<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > >::operator[](unsigned long) const /usr/include/c++/9/bits/stl_vector.h:1061
    #1 0x55e0adfa8ad6 in ORB_table_phi::cal_ST_Phi12_R(int const&, int const&, Numerical_Orbital_Lm const&, Numerical_Orbital_Lm const&, std::set<unsigned long, std::less<unsigned long>, std::allocator<unsigned long> > const&, double*, double*) const /root/abacus-develop/source/module_orbital/ORB_table_phi.cpp:334
    #2 0x55e0adedd6be in Center2_Orb::Orb11::init_radial_table(std::set<unsigned long, std::less<unsigned long>, std::allocator<unsigned long> > const&) /root/abacus-develop/source/src_lcao/center2_orb-orb11.cpp:74
    #3 0x55e0adee73f4 in Center2_Orb::Orb21::init_radial_table(std::set<unsigned long, std::less<unsigned long>, std::allocator<unsigned long> > const&) /root/abacus-develop/source/src_lcao/center2_orb-orb21.cpp:104
    #4 0x55e0ae1fe111 in Exx_Abfs::Matrix_Orbs21::init_radial_table(std::map<unsigned long, std::map<unsigned long, std::set<double, std::less<double>, std::allocator<double> >, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, std::set<double, std::less<double>, std::allocator<double> > > > >, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, std::map<unsigned long, std::set<double, std::less<double>, std::allocator<double> >, std::less<unsigned long>, std::allocator<std::pair<unsigned long const, std::set<double, std::less<double>, std::allocator<double> > > > > > > > const&) /root/abacus-develop/source/src_ri/exx_abfs-matrix_orbs21.cpp:179
    #5 0x55e0ae30ee94 in Exx_Abfs::PCA::cal_PCA(std::vector<std::vector<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> >, std::allocator<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> > > >, std::allocator<std::vector<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> >, std::allocator<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> > > > > > const&, std::vector<std::vector<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> >, std::allocator<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> > > >, std::allocator<std::vector<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> >, std::allocator<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> > > > > > const&, double) /root/abacus-develop/source/src_ri/exx_abfs-pca.cpp:58
    #6 0x55e0ae157b86 in Exx_Abfs::Construct_Orbs::pca(std::vector<std::vector<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> >, std::allocator<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> > > >, std::allocator<std::vector<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> >, std::allocator<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> > > > > > const&, std::vector<std::vector<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> >, std::allocator<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> > > >, std::allocator<std::vector<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> >, std::allocator<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> > > > > > const&, double, double) /root/abacus-develop/source/src_ri/exx_abfs-construct_orbs.cpp:281
    #7 0x55e0ae155f76 in Exx_Abfs::Construct_Orbs::abfs_same_atom(std::vector<std::vector<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> >, std::allocator<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> > > >, std::allocator<std::vector<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> >, std::allocator<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> > > > > > const&, double, double) /root/abacus-develop/source/src_ri/exx_abfs-construct_orbs.cpp:102
    #8 0x55e0adaae38b in Exx_LRI<std::complex<double> >::init(int const&) /root/abacus-develop/source/module_ri/Exx_LRI.hpp:60
    #9 0x55e0ada96544 in ModuleESolver::ESolver_KS_LCAO::Init(Input&, UnitCell&) /root/abacus-develop/source/module_esolver/esolver_ks_lcao.cpp:132
    #10 0x55e0ad8f0f5e in Driver::driver_run() /root/abacus-develop/source/driver_run.cpp:46
    #11 0x55e0ad8ee474 in Driver::atomic_world() /root/abacus-develop/source/driver.cpp:95
    #12 0x55e0ad8ed140 in Driver::init() /root/abacus-develop/source/driver.cpp:35
    #13 0x55e0ad3a9624 in main /root/abacus-develop/source/main.cpp:15
    #14 0x7f4bd1b1a082 in __libc_start_main ../csu/libc-start.c:308
    #15 0x55e0ad3a951d in _start (/usr/local/bin/abacus+0x2c051d)

0x612000041ba8 is located 24 bytes to the left of 288-byte region [0x612000041bc0,0x612000041ce0)
allocated by thread T0 here:
    #0 0x7f4bd3883587 in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cc:104
    #1 0x55e0ad402aef in __gnu_cxx::new_allocator<std::vector<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > > >::allocate(unsigned long, void const*) /usr/include/c++/9/ext/new_allocator.h:114
    #2 0x55e0ad401fde in std::allocator_traits<std::allocator<std::vector<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > > > >::allocate(std::allocator<std::vector<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > > >&, unsigned long) /usr/include/c++/9/bits/alloc_traits.h:443
    #3 0x55e0ad4012d7 in std::_Vector_base<std::vector<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > >, std::allocator<std::vector<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > > > >::_M_allocate(unsigned long) /usr/include/c++/9/bits/stl_vector.h:343
    #4 0x55e0ad3ff077 in std::vector<std::vector<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > >, std::allocator<std::vector<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > > > >::_M_default_append(unsigned long) /usr/include/c++/9/bits/vector.tcc:635
    #5 0x55e0ad3fe9fa in std::vector<std::vector<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > >, std::allocator<std::vector<std::vector<double, std::allocator<double> >, std::allocator<std::vector<double, std::allocator<double> > > > > >::resize(unsigned long) /usr/include/c++/9/bits/stl_vector.h:937
    #6 0x55e0ad3fcf74 in ModuleBase::Sph_Bessel_Recursive::D2::cal_jlx_0(int, unsigned long, unsigned long) /root/abacus-develop/source/module_base/sph_bessel_recursive-d2.cpp:41
    #7 0x55e0ad3fce05 in ModuleBase::Sph_Bessel_Recursive::D2::cal_jlx(int, unsigned long, unsigned long) /root/abacus-develop/source/module_base/sph_bessel_recursive-d2.cpp:32
    #8 0x55e0adfb095c in ORB_table_phi::init_Table_Spherical_Bessel(int, int, int&, int&, int const&, LCAO_Orbitals const&, Numerical_Nonlocal const*) /root/abacus-develop/source/module_orbital/ORB_table_phi.cpp:790
    #9 0x55e0ae1f9a98 in Exx_Abfs::Matrix_Orbs21::init(int, double, double) /root/abacus-develop/source/src_ri/exx_abfs-matrix_orbs21.cpp:40
    #10 0x55e0ae30ecaa in Exx_Abfs::PCA::cal_PCA(std::vector<std::vector<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> >, std::allocator<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> > > >, std::allocator<std::vector<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> >, std::allocator<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> > > > > > const&, std::vector<std::vector<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> >, std::allocator<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> > > >, std::allocator<std::vector<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> >, std::allocator<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> > > > > > const&, double) /root/abacus-develop/source/src_ri/exx_abfs-pca.cpp:48
    #11 0x55e0ae157b86 in Exx_Abfs::Construct_Orbs::pca(std::vector<std::vector<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> >, std::allocator<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> > > >, std::allocator<std::vector<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> >, std::allocator<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> > > > > > const&, std::vector<std::vector<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> >, std::allocator<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> > > >, std::allocator<std::vector<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> >, std::allocator<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> > > > > > const&, double, double) /root/abacus-develop/source/src_ri/exx_abfs-construct_orbs.cpp:281
    #12 0x55e0ae155f76 in Exx_Abfs::Construct_Orbs::abfs_same_atom(std::vector<std::vector<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> >, std::allocator<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> > > >, std::allocator<std::vector<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> >, std::allocator<std::vector<Numerical_Orbital_Lm, std::allocator<Numerical_Orbital_Lm> > > > > > const&, double, double) /root/abacus-develop/source/src_ri/exx_abfs-construct_orbs.cpp:102
    #13 0x55e0adaae38b in Exx_LRI<std::complex<double> >::init(int const&) /root/abacus-develop/source/module_ri/Exx_LRI.hpp:60
    #14 0x55e0ada96544 in ModuleESolver::ESolver_KS_LCAO::Init(Input&, UnitCell&) /root/abacus-develop/source/module_esolver/esolver_ks_lcao.cpp:132
    #15 0x55e0ad8f0f5e in Driver::driver_run() /root/abacus-develop/source/driver_run.cpp:46
    #16 0x55e0ad8ee474 in Driver::atomic_world() /root/abacus-develop/source/driver.cpp:95
    #17 0x55e0ad8ed140 in Driver::init() /root/abacus-develop/source/driver.cpp:35
    #18 0x55e0ad3a9624 in main /root/abacus-develop/source/main.cpp:15
    #19 0x7f4bd1b1a082 in __libc_start_main ../csu/libc-start.c:308
```

**Comment by xinyangd** [comment]
I also noticed this overflow problem when trying to debug the RPA interface. I will ask the contributor to see if he can fix it.

**Comment by maki49** [comment]
On my workstation (with Intel OneAPI 2021.3.0 MPI with OMP_NUM_THREADS=1) it gets stuck either. 
I will help follow it and talk to Peize Lin.

**Comment by mohanchen** [comment]
The parallel issue has been resolved. Some problems have been fixed in #2193.
